modules/lang-painless/src/test/java/org/elasticsearch/painless/ElvisTests.java

```diff
@@ -0,0 +1,116 @@
+        assertEquals(1, exec("int i = params.a ?: 2; return i", singletonMap("a", 1), true));
+
+        // Explicit casting
+        assertEquals(1, exec("return (Integer)(params.a ?: Integer.valueOf(1))"));
+        assertEquals(1, exec("return (Integer)(params.a ?: Integer.valueOf(2))", singletonMap("a", 1), true));
+        assertEquals(1, exec("return (int)(params.a ?: 1)"));
+        assertEquals(1, exec("return (int)(params.a ?: 2)", singletonMap("a", 1), true));
+
+        // Now some chains
+        assertEquals(1, exec("return params.a ?: params.a ?: 1"));
+        assertEquals(1, exec("return params.a ?: params.b ?: 'j'", singletonMap("b", 1), true));
+        assertEquals(1, exec("return params.a ?: params.b ?: 'j'", singletonMap("a", 1), true));
+
+        // Precedence
+        assertEquals(1, exec("return params.a ?: 2 + 2", singletonMap("a", 1), true));
+        assertEquals(4, exec("return params.a ?: 2 + 2"));
+        assertEquals(2, exec("return params.a + 1 ?: 2 + 2", singletonMap("a", 1), true)); // Yes, this is silly, but it should be valid
+
+        // Weird casts
+        assertEquals(1,     exec("int i = params.i;     String s = params.s; return s ?: i", singletonMap("i", 1), true));
+        assertEquals("str", exec("Integer i = params.i; String s = params.s; return s ?: i", singletonMap("s", "str"), true));
+
+        // Combining
+        assertEquals(2, exec("return (params.a ?: 0) + 1", singletonMap("a", 1), true));
+        assertEquals(1, exec("return (params.a ?: 0) + 1"));
+        assertEquals(2, exec("return (params.a ?: ['b': 10]).b + 1", singletonMap("a", singletonMap("b", 1)), true));
+        assertEquals(11, exec("return (params.a ?: ['b': 10]).b + 1"));
+    }
+
+    public void testWithNullSafeDereferences() {
+        assertEquals(1, exec("return params.a?.b ?: 1"));
+        assertEquals(1, exec("return params.a?.b ?: 2", singletonMap("a", singletonMap("b", 1)), true));
+    }
+
+    public void testLazy() {
+        assertEquals(1, exec("def fail() {throw new RuntimeException('test')} return params.a ?: fail()", singletonMap("a", 1), true));
+        Exception e = expectScriptThrows(RuntimeException.class, () ->
+            exec("def fail() {throw new RuntimeException('test')} return params.a ?: fail()"));
+        assertEquals(e.getMessage(), "test");
+    }
+
+    /**
+     * Checks that {@code a ?: b ?: c} is be parsed as {@code a ?: (b ?: c)} instead of {@code (a ?: b) ?: c} which is nice because the
+     * first one only needs one comparison if the {@code a} is non-null while the second one needs two.
+     */
+    public void testRightAssociative() {
+        /* Sadly this is a bit finicky about the output of the disassembly but I think it is worth having because it makes sure that
+         * the code generated for the elvis operator is as efficient as possible. */
+        String disassembled = Debugger.toString("params.a ?: params.b ?: params.c");
```

Comment:
I also played around with testing this by looking at the ANTLR AST or the Painless AST but I didn't really like either because they were more invasive and didn't test the thing I care about: that there is only one jump in the non-null path.